src/compiler/checker.ts

```diff
@@ -38932,6 +38952,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         const saveCurrentNode = currentNode;
         currentNode = node;
         instantiationCount = 0;
+        // #56013
+        if (node.kind === SyntaxKind.CallExpression && chooseOverloadRecursionLevel >= 0) {
+            let setOfNode: Set<Node> | undefined;
+            if (chooseOverloadRecursionLevel >= 0 && (setOfNode = chooseOverloadFlushNodesSignaturesReq[chooseOverloadRecursionLevel])) {
+                if (!setOfNode.has(node)) {
+                    /**
+                     * #56013
+                     * The following is an invariant condition beneath the above call to resolveSignature:
+                     * - getNodeLinks(node).resolvedSignature is either the global constant readonly value `resolvingSignature`
+                     * - or a calculated signature.
+                     * Therefore resetting must set to `resolvingSignature` - it means "not yet calculated".
+                     */
+                    getNodeLinks(node).resolvedSignature = resolvingSignature;

```

Comment:
I'm surprised the perf of this is OK, but it's likely because this solution is incomplete - this only partially purges the cached nodes; the _cached parameter symbol types_ on the parameter symbol links are still there! So while you'll get new signatures here, any parameters that get witnessed by multiple overload passes will still have their types locked down. Which isn't any worse than today, to be fair! It's a known issue with the signature resolution logic. IME, adding a full hierarchical cache to fully handle it and make speculative overload resolution totally unwindable was pretty bad for perf, but this approach, being partial, looks like it's OK. And if it still manages to fix a few usecases, that sounds pretty OK to me.

Reply:
Like, this feels odd, because it seems naive - it's basically resetting the resolved signature of *every* call expression we visit while looking at nested overload resolutions - in theory, only *some* calls should need to get reset - those which actually depend on the overload resolution in progress (which'd require a bit more tracing). I'd have figured over-invalidation like this'd manifest as terrible perf, but no - perf looks OK, broadly speaking, not more than a single percent diff.